packages/vendure-plugin-shipmate/src/api/shipmate.service.ts

```diff
@@ -0,0 +1,271 @@
+      'application/json';
+    this.httpService.axiosRef.defaults.headers.common['Content-Type'] =
+      'application/json';
+    this.eventBus.ofType(OrderPlacedEvent).subscribe(async ({ ctx, order }) => {
+      await this.setupRequestHeaders(ctx);
+      await this.createShipment(ctx, order);
+    });
+  }
+
+  async setupRequestHeaders(ctx: RequestContext) {
+    const shipmateConfig = await this.shipmateConfigService.getConfig(ctx);
+    if (!shipmateConfig) {
+      Logger.error(
+        `Shipmate credentials not configured for channel ${ctx.channel.code}`,
+        loggerCtx
+      );
+      return;
+    }
+    let shipmateTokenForTheCurrentChannel = this.tokens.get(
+      shipmateConfig.apiKey
+    );
+    if (!shipmateTokenForTheCurrentChannel) {
+      shipmateTokenForTheCurrentChannel = await this.getShipmentToken(
+        shipmateConfig.username,
+        shipmateConfig.password
+      );
+      this.tokens.set(shipmateConfig.apiKey, shipmateTokenForTheCurrentChannel);
+    }
+    this.httpService.axiosRef.defaults.headers.common[
+      SHIPMATE_TOKEN_HEADER_KEY
+    ] = shipmateTokenForTheCurrentChannel;
+    this.httpService.axiosRef.defaults.headers.common[
+      SHIPMATE_API_KEY_HEADER_KEY
+    ] = shipmateConfig.apiKey;
+  }
+
+  async getShipmentToken(
+    shipmateUsername: string,
+    shipmatePassword: string
+  ): Promise<string> {
+    const response = await firstValueFrom(
+      this.httpService.post<GetTokenRespose>(
+        `${this.config.shipmateApiUrl}/tokens`,
+        {
+          username: shipmateUsername,
+          password: shipmatePassword,
+        }
+      )
+    );
+    if (response.data.data?.token) {
+      Logger.info('Successfully authenticated with Shipmate API', loggerCtx);
+      return response.data.data.token;
+    } else {
+      Logger.error(response.data.message, loggerCtx);
+      throw new UserInputError(response.data.message);
+    }
+  }
+
+  async createShipment(ctx: RequestContext, order: Order) {
+    const shipmateReference = generatePublicId();
+    const payload = parseOrder(order, shipmateReference);
+    try {
+      const result = await firstValueFrom(
+        this.httpService.post<CreateShipmentResponse>(
+          `${this.config.shipmateApiUrl}/shipments`,
+          payload
+        )
+      );
+      Logger.info(result.data.message, loggerCtx);
+      await this.orderService.updateCustomFields(ctx, order.id, {
+        ...order.customFields,
+        shipmateReference,
+      });
+    } catch (error: any) {
+      const message = error.response.data.message;
+      Logger.error(message, loggerCtx);
+    }
+  }
+
+  async updateOrderState(payload: EventPayload, res: Response): Promise<void> {
+    const ctx = await this.createCtx(payload.auth_token);
+    if (!ctx) {
+      const message = `No registered ShipmentConfigEntity with this auth_token`;
+      Logger.error(message, loggerCtx);
+      res.status(HttpStatus.BAD_REQUEST).json({ message });
+      return;
```

Comment:
This should throw an Error, and let the controller handle HTTP response statusses. Logging is indeed a good practice though.

Reply:
I checked shipmate's api we don't really need to return 201. Any success code in the 200 series will do